Add circuit breaker to flake-report-creator

dhiller opened the pull request:
* Create a minimal circuitbreaker implementation that avoids stressing the Jenkins with requests in case the service returns 504 status indicating it is not healthy.
* To make the use case testable, create new interface BuildDataGetter for
dependency injection with default implementation. Extract method that
uses the interface while preserving the original. Use this interface for mock testing.
* Add test for parallel execution that assures that
the service is not called more often than wanted.

Note: the retry and jitter values were extended since the feedback was that the Jenkins services requires around 1-2 minutes to recover, therefore an extra minute to give it a bit more time.

TODOs:
- [x] [update the flake-report-creator image to use the new version](https://github.com/kubevirt/project-infra/pull/2414/files#diff-11d9dded96f1181674711c876b472f2e1eb22e1f430a666c3ac0d26a89e2fda0R9)

/cc @fossedihelm @xpivarc

dhiller commented:
@xpivarc @fossedihelm I've already assigned you for a first round of review to check whether this is a feasible approach, I need to update the images in case all is good.

/hold

fossedihelm commented on robots/pkg/circuitbreaker/circuitbreaker.go:
```diff
@@ -0,0 +1,81 @@
+//
+// [circuit breaker pattern]: https://martinfowler.com/bliki/CircuitBreaker.html
+type CircuitBreaker struct {
+	mutex      sync.Mutex
```
How about using `sync.RWMutex`? It will allow multiple reader at the same time, but only a single writer. https://pkg.go.dev/sync#RWMutex
And use only `RLock()` in `isOpenAndNotFeasibleForRetry()`?

dhiller replied:
TIL! Thanks!

fossedihelm commented on robots/pkg/circuitbreaker/circuitbreaker.go:
```diff
@@ -0,0 +1,81 @@
+	}
+}
+
+func (g *CircuitBreaker) isOpenAndNotFeasibleForRetry() (bool, error) {
+	g.mutex.Lock()
+	defer g.mutex.Unlock()
+	return g.open && !g.occurred.Add(g.retryAfter).Before(time.Now()), g.lastErr
+}
+
+func (g *CircuitBreaker) updateState(err error) {
+	g.mutex.Lock()
+	defer g.mutex.Unlock()
+	if err != nil {
+		g.open = true
+		g.occurred = time.Now()
```
Instead of using `occurred` field I would prefer to use something like `blockedUntil`. This will move the `Add(g.retryAfter)` in writing time, preventing to be calculated for every request if the circuit is open. I know I'm pretty punctilious right now. :grimacing:

dhiller replied:
Makes sense, thanks!

fossedihelm commented on robots/pkg/circuitbreaker/circuitbreaker.go:
```diff
@@ -0,0 +1,81 @@
+func (g *CircuitBreaker) updateState(err error) {
+	g.mutex.Lock()
+	defer g.mutex.Unlock()
+	if err != nil {
```
Shouldn't we open the circuit only in case of `http.StatusGatewayTimeout`?

fossedihelm replied:
If this is the case, I would try to make it more general; maybe adding a field containing a function that will be evaluated to determine when open the circuit. Just an idea :)

dhiller replied:
Good idea, makes sense!

fossedihelm reviewed (commented):
Thanks Daniel! It seems well done all things considered. I just have a doubt about when to open the circuit, because current implementation will open it at every error while in the description it should be done only in case of 504. Or am I missing something?
The others are details or implementation improvements.
Thanks

dhiller commented:
@fossedihelm all good points, thank you for the review!

fossedihelm commented on robots/pkg/circuitbreaker/circuitbreaker.go:
```diff
@@ -63,17 +67,17 @@ func (g *CircuitBreaker) WrapRetryableFunc(retryableFunc func() error) func() er
 	g.mutex.Lock()
 	defer g.mutex.Unlock()
-	if err != nil {
+	if err != nil && g.shouldOpen(err) {
```
Sorry for not noticing this earlier, but here probably we should check again if the circuit is open.
It should not happen, but let me try to explain my concern:
- A makes a request X at t<sup>0</sup>.
- B makes a request Y at t<sup>1</sup>.

Both A and B see the closed circuit, so the requests X,Y are executed.
Now let's assume that after a while(t<sup>3</sup>) X return 504; the circuit will be open with `blockedUntil` t<sup>5</sup>.
But after another while(t<sup>4</sup>) Y return 504 and the `blockedUntil` will be updated with t<sup>6</sup>.
Taking the reasoning to extremes, it may happen that the circuit remains open for twice the time it was set.
That is, the case where a 504 response to a request arrives exactly an instant before `blockedUntil`, as it will override it just a moment before it expires.
Checking the circuit before the write ensures that `blockedUntil` always refer to the first failure.
Sorry for the probably unclear comment. Let me know. Thanks

dhiller replied:
What you are describing to happen should be fine, since the failure of the second call indicates that the service under supervision is still not operating as it should, in which case the circuit breaker should extend the wait time for the latter occurrence. WDYT?

fossedihelm replied:
yes, but the service could be available after a `retryAfter` amount of time from the first occurrence of the error, but since we've updated the `blockUntil` just before the end of the first timeout we are going to wait for another `retryAfter` amount of time.

dhiller replied:
Added a check at the beginning of the update.

fossedihelm commented on robots/pkg/jenkins/builds_test.go:
```diff
@@ -142,6 +160,21 @@ var _ = Describe("builds.go", func() {
+				}()
+			}
+			wg.Wait()
+			Expect(buildDataGetter.GetCallCounter()).To(BeNumerically("<=", uint32(numberOfThreads*2)))
```
Can you explain why this is <= and not == ? Thanks

fossedihelm reviewed (commented):
Thanks @dhiller. Just another doubt. Sorry for the delay of this.

fossedihelm reviewed (commented):
Thanks, just one last nit that you are free to ignore :)

dhiller commented on robots/pkg/jenkins/builds_test.go:
```diff
@@ -0,0 +1,180 @@
+	BeforeEach(func() {
+		retryDelay = 150 * time.Millisecond
+		maxJitter = 10 * time.Millisecond
+		if retryDelay <= 0 {
```
Ah, thanks for the eagle eyes here. Was a refactoring leftover :+1:

dhiller commented:
@fossedihelm PTAL! (this time the final one hopefully :smile: )

fossedihelm reviewed (commented):
Thanks @dhiller Last comment and then I think that we are ready to go :) 
(sorry for being pedantic)

dhiller commented on robots/pkg/circuitbreaker/circuitbreaker.go:
```diff
@@ -69,10 +69,17 @@ func (g *CircuitBreaker) WrapRetryableFunc(retryableFunc func() error) func() er
 }
 
 func (g *CircuitBreaker) updateState(err error) {
+	if g.isRetryBlocked() {
+		return
+	}
```
IIUC what you are seeing is a read without a lock, could be fixed by adding a `RLock/RUnlock` call to `isRetryBlocked`. WDYT?

fossedihelm replied:
I prefer to make this specific read under the `Lock` context. Yes, it is a read, but it is done to allow or disallow a write.
Let me explain: if two process want to update, they both read and this can be done in parallel, so both are allowed to write. This is not a big deal since the subsequent write will be atomic (due to the `Lock`), but this does not prevent having multiple (very near) write, that could be avoided. WDYT?

dhiller replied:
Yes, that's a good point! Thanks!

dhiller commented:
/unhold

fossedihelm reviewed (approved):
Thanks for your patience!

dhiller commented:
> Thanks for your patience!

Thank you for your thorough review. I've updated the image sha in the script, would you mind taking another look again?

dhiller commented:
^^ @fossedihelm

fossedihelm commented:
/lgtm
You can check the todo in the description :slightly_smiling_face:

dhiller commented:
/hold something is wrong here

dhiller commented:
> /hold something is wrong here

/unhold I've added some logging of circuitbreaker state, replaced the default retry BackOff with a FixedDelay, and fixed one test (actually I've split into two scenarios)

fossedihelm commented:
/lgtm
Thanks

dhiller commented:
/unhold

dhiller commented:
/approve
